--- a/edit.c
+++ b/edit.c
@@ -91,7 +91,8 @@
 static int x_do_comment(char *, ssize_t, ssize_t *);
 static void x_print_expansions(int, char * const *, bool);
 static int x_cf_glob(int *, const char *, int, int, int *, int *, char ***);
-static size_t x_longest_prefix(int, char * const *);
+static size_t x_longest_prefix(int, char * const *, bool);
+static bool x_prefix_equal(const char *, const char *, size_t, bool);
 static void x_glob_hlp_add_qchar(char *);
 static char *x_glob_hlp_tilde_and_rem_qchar(char *, bool);
 static size_t x_basename(const char *, const char *);
@@ -255,7 +256,7 @@
 	 * case, we want to omit the directory name)
 	 */
 	if (!is_command &&
-	    (prefix_len = x_longest_prefix(nwords, words)) > 0) {
+	    (prefix_len = x_longest_prefix(nwords, words, false)) > 0) {
 		int i;
 
 		/* Special case for 1 match (prefix is whole word) */
@@ -423,7 +424,7 @@
 		/* this always results in at most one match */
 		nwords = 0;
 	}
-	expand(yylval.cp, &w, nwords);
+	expand_icase(yylval.cp, &w, nwords);
 	XPput(w, NULL);
 	words = (char **)XPclose(w);
 
@@ -702,8 +703,22 @@
 /*
  * Find longest common prefix
  */
+static bool
+x_prefix_equal(const char *a, const char *b, size_t len, bool ignorecase)
+{
+	while (len--) {
+		if (*a != *b && (!ignorecase || !ctype(*a, C_ALPHA) ||
+		    !ctype(*b, C_ALPHA) ||
+		    ksh_tolower(*a) != ksh_tolower(*b)))
+			return (false);
+		++a;
+		++b;
+	}
+	return (true);
+}
+
 static size_t
-x_longest_prefix(int nwords, char * const * words)
+x_longest_prefix(int nwords, char * const * words, bool ignorecase)
 {
 	int i;
 	size_t j, prefix_len;
@@ -715,7 +730,8 @@
 	prefix_len = strlen(words[0]);
 	for (i = 1; i < nwords; i++)
 		for (j = 0, p = words[i]; j < prefix_len; j++)
-			if (p[j] != words[0][j]) {
+			if (!x_prefix_equal(p + j, words[0] + j, 1,
+			    ignorecase)) {
 				prefix_len = j;
 				break;
 			}
@@ -782,7 +798,7 @@
 
 	ktwalk(&ts, tp);
 	while ((te = ktnext(&ts)))
-		if (gmatchx(te->name, pat, false)) {
+		if (gmatchx_icase(te->name, pat, false)) {
 			char *cp;
 
 			strdupx(cp, te->name, ATEMP);
@@ -829,7 +845,7 @@
 
 		oldsize = XPsize(*wp);
 		/* mark dirs */
-		glob_str(Xstring(xs, xp), wp, true);
+		glob_str(Xstring(xs, xp), wp, true, true);
 		newsize = XPsize(*wp);
 
 		/* Check that each match is executable... */
@@ -2795,7 +2811,7 @@
 		return;
 	}
 	olen = end - start;
-	nlen = x_longest_prefix(nwords, words);
+	nlen = x_longest_prefix(nwords, words, true);
 	if (nwords == 1) {
 		/*
 		 * always complete single matches;
@@ -2817,7 +2833,8 @@
 		 * longest prefix, implying the latter is at least
 		 * the same size (after unescaping)
 		 */
-		completed = !strncmp(words[0], unescaped, strlen(unescaped));
+		completed = x_prefix_equal(words[0], unescaped,
+		    strlen(unescaped), true);
 
 		afree(unescaped, ATEMP);
 	}
@@ -5446,7 +5463,7 @@
 		/* expanded = PRINT;	next call undo */
 	} else {
 		match = words[0];
-		match_len = x_longest_prefix(nwords, words);
+		match_len = x_longest_prefix(nwords, words, true);
 		/* next call will list completions */
 		expanded = COMPLETE;
 		is_unique = nwords == 1;
--- a/eval.c
+++ b/eval.c
@@ -80,13 +80,14 @@
 static int comsub(Expand *, const char *, int);
 static char *valsub(struct op *, Area *);
 static char *trimsub(char *, char *, int);
-static void glob(char *, XPtrV *, bool);
+static void expand0(const char *, XPtrV *, int, bool);
+static void glob(char *, XPtrV *, bool, bool);
 static void globit(XString *, char **, char *, XPtrV *, int);
 static const char *maybe_expand_tilde(const char *, XString *, char **, bool);
 #ifndef MKSH_NOPWNAM
 static char *homedir(char *);
 #endif
-static void alt_expand(XPtrV *, char *, char *, char *, int);
+static void alt_expand(XPtrV *, char *, char *, char *, int, bool);
 static int utflen(const char *) MKSH_A_PURE;
 static void utfincptr(const char *, mksh_ari_t *);
 
@@ -225,6 +226,26 @@
     /* DO* flags */
     int f)
 {
+	expand0(ccp, wp, f, false);
+}
+
+void
+expand_icase(const char *ccp, XPtrV *wp, int f)
+{
+	expand0(ccp, wp, f, true);
+}
+
+static void
+expand0(
+    /* input word */
+    const char *ccp,
+    /* output words */
+    XPtrV *wp,
+    /* DO* flags */
+    int f,
+    /* case-insensitive filename globbing */
+    bool ignorecase)
+{
 	int c = 0;
 	/* expansion type */
 	int type;
@@ -1072,9 +1093,10 @@
 					/* also does globbing */
 					alt_expand(wp, cp, cp,
 					    cp + Xlength(ds, (dp - 1)),
-					    fdo | (f & DOMARKDIRS));
+					    fdo | (f & DOMARKDIRS), ignorecase);
 				else if (fdo & DOGLOB)
-					glob(cp, wp, tobool(f & DOMARKDIRS));
+					glob(cp, wp, tobool(f & DOMARKDIRS),
+					    ignorecase);
 				else if ((f & DOPAT) || !(fdo & DOMAGIC))
 					XPput(*wp, cp);
 				else
@@ -1708,11 +1730,11 @@
 
 /* XXX cp not const 'cause slashes are temporarily replaced with NULs... */
 static void
-glob(char *cp, XPtrV *wp, bool markdirs)
+glob(char *cp, XPtrV *wp, bool markdirs, bool ignorecase)
 {
 	int oldsize = XPsize(*wp);
 
-	if (glob_str(cp, wp, markdirs) == 0)
+	if (glob_str(cp, wp, markdirs, ignorecase) == 0)
 		XPput(*wp, debunk(cp, cp, strlen(cp) + 1));
 	else
 		qsort(XPptrv(*wp) + oldsize, XPsize(*wp) - oldsize,
@@ -1723,20 +1745,22 @@
 #define GF_EXCHECK	BIT(0)		/* do existence check on file */
 #define GF_GLOBBED	BIT(1)		/* some globbing has been done */
 #define GF_MARKDIR	BIT(2)		/* add trailing / to directories */
+#define GF_ICASE	BIT(3)		/* case-insensitive matching */
 
 /*
  * Apply file globbing to cp and store the matching files in wp. Returns
  * the number of matches found.
  */
 int
-glob_str(char *cp, XPtrV *wp, bool markdirs)
+glob_str(char *cp, XPtrV *wp, bool markdirs, bool ignorecase)
 {
 	int oldsize = XPsize(*wp);
 	XString xs;
 	char *xp;
 
 	Xinit(xs, xp, 256, ATEMP);
-	globit(&xs, &xp, cp, wp, markdirs ? GF_MARKDIR : GF_NONE);
+	globit(&xs, &xp, cp, wp, (markdirs ? GF_MARKDIR : GF_NONE) |
+	    (ignorecase ? GF_ICASE : GF_NONE));
 	Xfree(xs, xp);
 
 	return (XPsize(*wp) - oldsize);
@@ -1855,14 +1879,15 @@
 				/* always ignore . and .. */
 				continue;
 			if ((*name == '.' && *sp != '.') ||
-			    !gmatchx(name, sp, true))
+			    !(check & GF_ICASE ? gmatchx_icase(name, sp, true) :
+			    gmatchx(name, sp, true)))
 				continue;
 
 			len = strlen(d->d_name) + 1;
 			XcheckN(*xs, xp, len);
 			memcpy(xp, name, len);
 			*xpp = xp + len - 1;
-			globit(xs, xpp, np, wp, (check & GF_MARKDIR) |
+			globit(xs, xpp, np, wp, (check & (GF_MARKDIR | GF_ICASE)) |
 			    GF_GLOBBED | (np ? GF_EXCHECK : GF_NONE));
 			xp = Xstring(*xs, xp) + prefix_len;
 		}
@@ -2009,7 +2034,8 @@
 #endif
 
 static void
-alt_expand(XPtrV *wp, char *start, char *exp_start, char *end, int fdo)
+alt_expand(XPtrV *wp, char *start, char *exp_start, char *end, int fdo,
+    bool ignorecase)
 {
 	unsigned int count = 0;
 	char *brace_start, *brace_end, *comma = NULL;
@@ -2046,14 +2072,14 @@
 		 * expansion. }
 		 */
 		if (fdo & DOGLOB)
-			glob(start, wp, tobool(fdo & DOMARKDIRS));
+			glob(start, wp, tobool(fdo & DOMARKDIRS), ignorecase);
 		else
 			XPput(*wp, debunk(start, start, end - start));
 		return;
 	}
 	brace_end = p;
 	if (!comma) {
-		alt_expand(wp, start, brace_end, end, fdo);
+		alt_expand(wp, start, brace_end, end, fdo, ignorecase);
 		return;
 	}
 
@@ -2082,7 +2108,7 @@
 				memcpy(news + l1 + l2, brace_end, l3);
 				news[l1 + l2 + l3] = '\0';
 				alt_expand(wp, news, news + l1,
-				    news + l1 + l2 + l3, fdo);
+				    news + l1 + l2 + l3, fdo, ignorecase);
 				field_start = p + 1;
 			}
 		}
--- a/misc.c
+++ b/misc.c
@@ -51,9 +51,10 @@
     const unsigned char *, bool) MKSH_A_PURE;
 static int do_gmatch(const unsigned char *, const unsigned char *,
     const unsigned char *, const unsigned char *,
-    const unsigned char *) MKSH_A_PURE;
-static const unsigned char *gmatch_cclass(const unsigned char *, unsigned char)
-    MKSH_A_PURE;
+    const unsigned char *, bool) MKSH_A_PURE;
+static const unsigned char *gmatch_cclass(const unsigned char *, unsigned char,
+    bool) MKSH_A_PURE;
+static bool gmatch_same(unsigned char, unsigned char, bool) MKSH_A_PURE;
 #ifdef KSH_CHVT_CODE
 static void chvt(const Getopt *);
 #endif
@@ -676,8 +677,15 @@
  * Match a pattern as in sh(1).
  * pattern character are prefixed with MAGIC by expand.
  */
-int
-gmatchx(const char *s, const char *p, bool isfile)
+static bool
+gmatch_same(unsigned char a, unsigned char b, bool ignorecase)
+{
+	return (a == b || (ignorecase && ctype(a, C_ALPHA) &&
+	    ctype(b, C_ALPHA) && ksh_tolower(a) == ksh_tolower(b)));
+}
+
+static int
+gmatchx_internal(const char *s, const char *p, bool isfile, bool ignorecase)
 {
 	const char *se, *pe;
 	char *pnew;
@@ -696,7 +704,13 @@
 		char tbuf[64];
 		char *t = len <= sizeof(tbuf) ? tbuf : alloc(len, ATEMP);
 		debunk(t, p, len);
-		return (!strcmp(t, s));
+		if (!ignorecase)
+			return (!strcmp(t, s));
+		while (*t && *s && gmatch_same(*t, *s, true)) {
+			++t;
+			++s;
+		}
+		return (*t == *s);
 	}
 	se = strnul(s);
 
@@ -709,11 +723,23 @@
 
 	rv = do_gmatch((const unsigned char *)s, (const unsigned char *)se,
 	    (const unsigned char *)pnew, (const unsigned char *)pe,
-	    (const unsigned char *)s);
+	    (const unsigned char *)s, ignorecase);
 	afree(pnew, ATEMP);
 	return (rv);
 }
 
+int
+gmatchx(const char *s, const char *p, bool isfile)
+{
+	return (gmatchx_internal(s, p, isfile, false));
+}
+
+int
+gmatchx_icase(const char *s, const char *p, bool isfile)
+{
+	return (gmatchx_internal(s, p, isfile, true));
+}
+
 /**
  * Returns if p is a syntacticly correct globbing pattern, false
  * if it contains no pattern characters or if there is a syntax error.
@@ -816,7 +842,7 @@
 static int
 do_gmatch(const unsigned char *s, const unsigned char *se,
     const unsigned char *p, const unsigned char *pe,
-    const unsigned char *smin)
+    const unsigned char *smin, bool ignorecase)
 {
 	unsigned char sc, pc, sl = 0;
 	const unsigned char *prest, *psub, *pnext;
@@ -831,7 +857,7 @@
 		sc = s < se ? *s : '\0';
 		s++;
 		if (!ISMAGIC(pc)) {
-			if (sc != pc)
+			if (!gmatch_same(sc, pc, ignorecase))
 				return (0);
 			sl = sc;
 			continue;
@@ -861,7 +887,8 @@
 				/* neither */
 				return (0);
 			}
-			if (sc == 0 || (p = gmatch_cclass(p, sc)) == NULL)
+			if (sc == 0 ||
+			    (p = gmatch_cclass(p, sc, ignorecase)) == NULL)
 				return (0);
 			break;
 
@@ -879,7 +906,7 @@
 				return (1);
 			s--;
 			do {
-				if (do_gmatch(s, se, p, pe, smin))
+				if (do_gmatch(s, se, p, pe, smin, ignorecase))
 					return (1);
 			} while (s++ < se);
 			return (0);
@@ -898,15 +925,17 @@
 			s--;
 			/* take care of zero matches */
 			if (ord(p[-1]) == (0x80 | ORD('*')) &&
-			    do_gmatch(s, se, prest, pe, smin))
+			    do_gmatch(s, se, prest, pe, smin, ignorecase))
 				return (1);
 			for (psub = p; ; psub = pnext) {
 				pnext = pat_scan(psub, pe, true);
 				for (srest = s; srest <= se; srest++) {
-					if (do_gmatch(s, srest, psub, pnext - 2, smin) &&
-					    (do_gmatch(srest, se, prest, pe, smin) ||
+					if (do_gmatch(s, srest, psub, pnext - 2, smin,
+					    ignorecase) && (do_gmatch(srest, se, prest, pe,
+					    smin, ignorecase) ||
 					    (s != srest &&
-					    do_gmatch(srest, se, p - 2, pe, smin))))
+					    do_gmatch(srest, se, p - 2, pe, smin,
+					    ignorecase))))
 						return (1);
 				}
 				if (pnext == prest)
@@ -925,14 +954,15 @@
 			s--;
 			/* Take care of zero matches */
 			if (ord(p[-1]) == (0x80 | ORD('?')) &&
-			    do_gmatch(s, se, prest, pe, smin))
+			    do_gmatch(s, se, prest, pe, smin, ignorecase))
 				return (1);
 			for (psub = p; ; psub = pnext) {
 				pnext = pat_scan(psub, pe, true);
 				srest = prest == pe ? se : s;
 				for (; srest <= se; srest++) {
-					if (do_gmatch(s, srest, psub, pnext - 2, smin) &&
-					    do_gmatch(srest, se, prest, pe, smin))
+					if (do_gmatch(s, srest, psub, pnext - 2, smin,
+					    ignorecase) && do_gmatch(srest, se, prest, pe,
+					    smin, ignorecase))
 						return (1);
 				}
 				if (pnext == prest)
@@ -951,7 +981,7 @@
 				for (psub = p; ; psub = pnext) {
 					pnext = pat_scan(psub, pe, true);
 					if (do_gmatch(s, srest, psub,
-					    pnext - 2, smin)) {
+					    pnext - 2, smin, ignorecase)) {
 						matched = 1;
 						break;
 					}
@@ -959,13 +989,14 @@
 						break;
 				}
 				if (!matched &&
-				    do_gmatch(srest, se, prest, pe, smin))
+				    do_gmatch(srest, se, prest, pe, smin,
+				    ignorecase))
 					return (1);
 			}
 			return (0);
 
 		default:
-			if (sc != p[-1])
+			if (!gmatch_same(sc, p[-1], ignorecase))
 				return (0);
 			break;
 		}
@@ -1009,7 +1040,7 @@
 };
 
 static const unsigned char *
-gmatch_cclass(const unsigned char *pat, unsigned char sc)
+gmatch_cclass(const unsigned char *pat, unsigned char sc, bool ignorecase)
 {
 	unsigned char c, subc, lc;
 	const unsigned char *p = pat, *s;
@@ -1074,8 +1105,11 @@
 					while (cls->name)
 						if (!strcmp(subp, cls->name)) {
 							/* found, match? */
-							if (ctype(sc,
-							    cls->value))
+							if (ctype(sc, cls->value) ||
+							    (ignorecase &&
+							    (cls->value == C_LOWER ||
+							    cls->value == C_UPPER) &&
+							    ctype(sc, C_ALPHA)))
 								found = true;
 							/* break either way */
 							break;
@@ -1099,7 +1133,7 @@
 				/* 0x80|' ' is plain (...) */
 				if ((c &= 0x7F) != ' ') {
 					/* check single match NOW */
-					if (sc == c)
+					if (gmatch_same(sc, c, ignorecase))
 						found = true;
 					/* next character is (...) */
 				}
@@ -1111,7 +1145,7 @@
 		    /* not terminating bracket? */
 		    (!ISMAGIC(p[2]) || ord(p[3]) != ORD(']')))) {
 			/* no, check single match */
-			if (sc == c)
+			if (gmatch_same(sc, c, ignorecase))
 				/* note: sc is never NUL */
 				found = true;
 			/* do the next "first" character */
@@ -1162,7 +1196,7 @@
 					/* oops, not a range */
 
 					/* match single previous char */
-					if (lc && (sc == lc))
+					if (lc && gmatch_same(sc, lc, ignorecase))
 						found = true;
 					/* match hyphen-minus */
 					if (ord(sc) == ORD('-'))
@@ -1191,10 +1225,19 @@
 			}
 		}
 		/* now do the actual range match check */
-		if (lc != 0 /* && c != 0 */ &&
-		    asciibetical(lc) <= asciibetical(sc) &&
-		    asciibetical(sc) <= asciibetical(c))
-			found = true;
+		if (lc != 0 /* && c != 0 */) {
+			unsigned char rsc = sc, rlc = lc, rc = c;
+
+			if (ignorecase && ctype(rsc, C_ALPHA) &&
+			    ctype(rlc, C_ALPHA) && ctype(rc, C_ALPHA)) {
+				rsc = ksh_tolower(rsc);
+				rlc = ksh_tolower(rlc);
+				rc = ksh_tolower(rc);
+			}
+			if (asciibetical(rlc) <= asciibetical(rsc) &&
+			    asciibetical(rsc) <= asciibetical(rc))
+				found = true;
+		}
 		/* forced next character? */
 		if (subc) {
 			c = subc;
--- a/sh.h
+++ b/sh.h
@@ -2420,7 +2420,8 @@
 char *evalonestr(const char *cp, int);
 char *debunk(char *, const char *, size_t);
 void expand(const char *, XPtrV *, int);
-int glob_str(char *, XPtrV *, bool);
+void expand_icase(const char *, XPtrV *, int);
+int glob_str(char *, XPtrV *, bool, bool);
 char *do_tilde(char *);
 /* exec.c */
 int execute(struct op * volatile, volatile int, volatile int * volatile);
@@ -2644,6 +2645,7 @@
 int parse_args(const char **, int, bool *);
 int getn(const char *, int *);
 int gmatchx(const char *, const char *, bool);
+int gmatchx_icase(const char *, const char *, bool);
 bool has_globbing(const char *) MKSH_A_PURE;
 int ascstrcmp(const void *, const void *) MKSH_A_PURE;
 int ascpstrcmp(const void *, const void *) MKSH_A_PURE;
